File: 518-coin-change-2/518-coin-change-2.java
class Solution {
    public int change(int amount, int[] coins) {        
        
        int[] ways = new int[amount+1]; ways[0] = 1;
        
        for(int c : coins)
          for(int i = 1; i <= amount; i++)
              if (i - c >= 0) ways[i] +=  ways[i-c];
          
        return ways[amount];
    }
}

File: 518-coin-change-2/518-coin-change-2.py
class Solution:
    def change(self, amount: int, coins: List[int]) -> int:
        N = len(coins)
        dp = [[0]*(amount+1) for _ in range(N + 1)]
        dp[0][0] = 1
        
        for amt in range(amount + 1):
            for i in range(1, N + 1):
                dp[i][amt] = dp[i-1][amt]
                if amt - coins[i-1] >= 0:
                    dp[i][amt] += dp[i][amt - coins[i-1]]
            
        return dp[N][amount]